### PR TITLE
Refactor string messages to objects

### PR DESCRIPTION
**Pull request by mirpo**

### CODE REVIEW

File: src/NetworkNode.js [hunk 1]
```diff
@@ -8,11 +9,16 @@ module.exports = class NetworkNode extends Node {
...
+        const dataMessage = new DataMessage()
+        dataMessage.setStreamId(streamId)
+        dataMessage.setPayload(content)
+
+        this.onDataReceived(dataMessage)
     }
 
     addMessageListener(cb) {
-        this.on(Node.events.MESSAGE_RECEIVED, (streamId, content) => cb(streamId, 0, content))
+        this.node.on(Node.events.MESSAGE_RECEIVED, (dataMessage) => cb(dataMessage.getStreamId(), 0, dataMessage.getPayload()))
```
Comment (harbu) [inline src/NetworkNode.js]: `this.node.on` should be `this.on`

File: src/helpers/MessageEncoder.js [hunk 2]
```diff
@@ -22,15 +27,36 @@ const encode = (type, data) => {
...
+                version, code, source, data
+            })
+        default:
+            return Object.assign(new BasicMessage(), {
+                version, code, source, data
+            })
     }
 }
 
+// const createMessage = (code) => new BasicMessage(CURRENT_VERSION, code)
```
Comment (harbu) [inline src/helpers/MessageEncoder.js]: Just remove commented out code?

File: src/logic/Node.js [hunk 3]
```diff
@@ -41,17 +42,10 @@ class Node extends EventEmitter {
...
-        })
-        this.protocols.nodeToNode.on(NodeToNode.events.DATA_RECEIVED, ({ streamId, data }) => this.onDataReceived(streamId, data))
-        this.protocols.nodeToNode.on(NodeToNode.events.SUBSCRIBE_REQUEST, ({ streamId, sender }) => {
-            this.onSubscribeRequest(streamId, sender)
-        })
-
-        this.protocols.nodeToNode.on(NodeToNode.events.UNSUBSCRIBE_REQUEST, ({ streamId, sender }) => {
-            this.onUnsubscribeRequest(streamId, sender)
-        })
+        this.protocols.trackerNode.on(TrackerNode.events.STREAM_INFO_RECEIVED, (streamMessage) => this.addKnownStreams(streamMessage))
```
Comment (harbu) [inline src/logic/Node.js:45]: 👍

Comment (harbu) [reply src/logic/Node.js:45]: Good that parameter names match class types, makes code more readable.

File: src/helpers/MessageEncoder.js [hunk 4]
```diff
@@ -22,15 +27,36 @@ const encode = (type, data) => {
...
+        case msgTypes.STREAM:
+            return Object.assign(new StreamMessage(), {
+                version, code, source, data
+            })
+        case msgTypes.DATA:
+            return Object.assign(new DataMessage(), {
+                version, code, source, data
+            })
+        case msgTypes.SUBSCRIBE:
+        case msgTypes.UNSUBSCRIBE:
```
Comment (harbu) [inline src/helpers/MessageEncoder.js:47]: Do you think we need separate `UnsubscribeMessage` in future?

File: src/helpers/MessageEncoder.js [hunk 5]
```diff
@@ -22,15 +27,36 @@ const encode = (type, data) => {
...
+const decode = (source, message) => {
     const { version, code, data } = JSON.parse(message)
-    return {
-        version,
-        code,
-        data
+
+    switch (code) {
+        case msgTypes.STATUS:
+            return Object.assign(new StatusMessage(), {
```
Comment (harbu) [inline src/helpers/MessageEncoder.js:35]: Why not constructor arguments?

File: src/logic/Node.js [hunk 6]
```diff
@@ -185,7 +185,14 @@ class Node extends EventEmitter {
 
     _handleBufferedMessages(streamId) {
         this.messageBuffer.popAll(streamId)
-            .forEach((data) => this.onDataReceived(streamId, data))
+            .forEach((payload) => {
+                const dataMessage = new DataMessage()
+                dataMessage.setStreamId(streamId)
+                dataMessage.setPayload(payload)
+
+                // bad idea to call events directly
```
Comment (harbu) [inline src/logic/Node.js]: Could you put `TODO:` in front? Makes easier to `grep` these sorts of things later on. Regarding invoking event handler directly, I guess we could wrap the thing we really want to do into a separate function that both the event handler and this function invokes.

File: src/messages/DataMessage.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+module.exports = class DataMessage extends BasicMessage {
+    getStreamId() {
+        return this.data[0]
+    }
+
+    setStreamId(streamId) {
+        this.data[0] = streamId
+    }
+
+    getPayload() {
```
Comment (harbu) [inline src/messages/DataMessage.js:12]: I like this naming. `Data` for the protocol message stuff and `payload` for the actual content in the message. Let's try to use this separation in all code. 👍

File: src/messages/SubscribeMessage.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+module.exports = class StreamMessage extends BasicMessage {
+    getStreamId() {
+        return this.data
+    }
+
+    setStreamId(streamId) {
+        this.data = streamId
+    }
+
+    getSender() {
```
Comment (harbu) [inline src/messages/SubscribeMessage.js:12]: Should just use `getSource()` as elsewhere?

File: src/connection/Libp2pEndpoint.js [hunk 9]
```diff
@@ -64,10 +65,7 @@ class Libp2pEndpoint extends EventEmitter {
                 pull.drain((message) => {
                     debug('received from peer %s message with data "%s"', sender instanceof PeerInfo ? getAddress(sender) : '', message)
 
-                    this.emit(events.MESSAGE_RECEIVED, {
-                        sender,
-                        message
-                    })
+                    this.emit(events.MESSAGE_RECEIVED, encoder.decode(sender, message))
```
Comment (ptesavol) [inline src/connection/Libp2pEndpoint.js]: Here message is being decoded on the connection layer. Messages should be decoded on the protocol layer (what if there is a protocol that wants the raw data? we can never know)

File: src/helpers/MessageEncoder.js (hunk 5 above)
Comment (ptesavol) [inline src/helpers/MessageEncoder.js:35]: Wouldn't it be better to use constructors instead?

File: src/logic/Node.js [hunk 10]
```diff
@@ -185,7 +185,14 @@ class Node extends EventEmitter {
 
     _handleBufferedMessages(streamId) {
         this.messageBuffer.popAll(streamId)
-            .forEach((data) => this.onDataReceived(streamId, data))
+            .forEach((payload) => {
+                const dataMessage = new DataMessage()
+                dataMessage.setStreamId(streamId)
```
Comment (ptesavol) [inline src/logic/Node.js:190]: Would be better to use constructor parameters

File: src/protocol/TrackerServer.js [hunk 11]
```diff
@@ -40,33 +40,24 @@ class TrackerServer extends EventEmitter {
...
         if (!isTracker(peer)) {
             this.emit(events.NODE_CONNECTED, peer)
         }
     }
 
-    onMessageReceived(peer, message) {
-        const { code, data } = encoder.decode(message)
-
-        switch (code) {
+    onMessageReceived(message) {
```
Comment (ptesavol) [inline src/protocol/TrackerServer.js:49]: onMessageReceived() should receive raw data and do the decoding

**Review by ptesavol — CHANGES_REQUESTED** [review]
Protocol layer should receive raw data, because we cannot make assumptions about the format of data a protocol wants to receive from the network. Message objects should be initialized using constructors instead of the ways that are used now. (there is also 4 comments in the code, most of them concern these 2 matters=

File: test/integration/libp2pendpoint.test.js [hunk 12]
```diff
@@ -1,42 +1,46 @@
...
+
+        for (let i = 0; i < MAX; i++) {
+            expect(nodes[i].protocols.nodeToNode.endpoint.getPeers().length).toEqual(2)
+        }
 
         promises = []
         for (let i = 0; i < MAX; i++) {
-            // eslint-disable-next-line no-await-in-loop
-            promises.push(await endpoints[i].stop(console.log(`closing ${i} endpoint`)))
+            promises.push(nodes[i].protocols.nodeToNode.endpoint.stop(() => console.log(`closing ${i} endpoint`)))
```
Comment (harbu) [inline test/integration/libp2pendpoint.test.js]: should we use `node.stop()`?

File: test/integration/publisher.test.js [hunk 13]
```diff
@@ -1,39 +0,0 @@
-const assert = require('assert')
```
Comment (harbu) [inline test/integration/publisher.test.js:1]: This test was deleted. No longer needed?

**Review by harbu — COMMENTED** [review]
Looks good. Few questsions; See comments

**Pull request merged**